README.txt
WELCOME:
--------
Welcome to CandyDollDB, a web based way of browsing, updating and
enjoying your CandyDollTV-collection.



DATE & VERSION:
---------------
2012-07-08, v1.7



NEWS:
-----
It took some time, but it was well worth the wait. This update brings
multi-lingual readiness including a Dutch translation.

Tagging, tags and tag-based search capabilities have been added, with
all known sets tagged to the best of our knowledge.

Would you like to translate CandyDollDB into your own language? Do you
want to help in tagging CD's sets? Feel free to contact me (see CONTACT). 



SYSTEM REQUIREMENTS:
--------------------
Computer running
 * PHP 5.3
 * MySQL 5.1
 * SMTP-mail server (optional)



FRESH INSTALL:
--------------
 * Extract the downloaded archive into a directory which is accessible 
through your web server.
 * Make sure the user running the web server has read-write access to 
the CandyDollDB root- and cache directories.
 * Make sure the user running the web server has read-access to your 
CandyDoll-collection.
 * Using your browser, navigate to setup.php and follow the on screen 
instructions.
 * NOTE: Any previous CandyDollDB-database will be dropped.
 * After setup is complete, log in and click Features -> Process XML.
 * Import your images and videos per model, per set or on the commandline.
 
  

HISTORY:
--------
2012-07-08 1.7
		Added multi-lingual support including Dutch translation;
		Added user-specific dateformats;
		Added user-specific choice between table- and thumbnail views;
		Added the most recent model- and set updates;
		Added tagging of models, sets, images and videos;
		Added search capability based on tagged items, thanks to ErikG;
		Added autosuggest, import and export for tags;
		Added paging for search- and dirty set-pages;
		Added commandline error-output;
		Added option to export only tagged images/videos;
		Extended information for each model on index;
		Included an 'offline' copy of jQuery;
		Fixed minor GUI-issues, lots of code clean-up;
		Fixed the skewed portrait-only thumbnail-bug;
		Implemented natural sorting for sets;

2012-02-26 1.6
		Added support for a custom databasename;
		Added the most recent releasedates from the official CD-site;
		Added model thumbnails on main page;
		Added the most probable SvetlanaT06 release-date, thanks to V@RTEX;
		Added ColorBox including slideshow feature, removed SlimBox;
		Added a 'Remarks' field to Models;
		Added a delete-on-overlay feature for deleting cacheimages;
		Added a split pics- and videolist of dirty sets on main page;
		Added functionality for cleaning both cachetable and -folder;
		Fixed the stretching of thumbnail images on portrait-only sets;
		Tweaked the set import to decrease number of dirty sets;
		Tweaked timeout setting for lengthy operations;
		Updated the PHPMailer class;
		Centralized the handling of querystring variables; 
		Changed the inner workings of the random-image grabber;
		Minor cosmetic surgery, bugfixes and code clean up; 

2012-01-28 1.5
		Added a new lay-out and design (Vika is such a beauty, isn't she?);
		Added an admin-panel for accessing functionality found nowhere else;
		Added an ul-li-fied navigation menu;
		Added the CANDYDOLLDB_VERSION constant;
		Added the most recent updates;
		Added physical cache-folder cleaning;
		Refined cache-functionality (auto-delete when appropriate);
		Fixed CLI progress-indicator issue;
		Fixed TimeZone issue - thanks to gothicpinku;
		Fixed the centering of erroroverlay issue;
		Fixed a low-memory issue when downloading large zipfiles;
		Fixed some other minor bugs;
		Removed the alternate style directives (see config.php and setup.php); 		

2012-01-04 1.4
		Added support for a blank database password - thanks to Kasimi;
		Added OS-detection for image- and video-paths during setup;
		Added auto-redirect after session-expiration;
		Added support for requests through HTTPS without unsafe content warnings;
		Added a more elaborate caching mechanism, for models, indexes, sets,
		 videos and images;
		Added thumbnailed set overview instead of a simple table; 
		Added a multi-download form for custom-built zip downloads;
		Added XML-export functionality for models and sets;
		Added a very big list of model birthdates - thanks to CandyGirl;
		Added a very big list of releasedates - thanks to Archive.org;
		Fixed a potential security hole: exit after header(location); 
		Tons of code clean-up and minor cosmetic surgery;

2011-09-25 1.3
		Added an automatic index generator;
		Added a caching mechanism for this generator;
		Added lots of new models and release-dates;
		Added convenient filenames for zip-exports;
		Added dates and -search capability on dirty sets;
		Added an alternate style (i.e. optional override);
		Added a password generator;
		Minor bugfix in importscript;
		Minor UI-fix in focussing text inputs;

2011-07-21 1.2
		SQL-user bug fixed in setup-script;
		Multiple 'undefined'-bugs in imports fixed;

2011-07-18 1.1
		Added support for multiple image- or videodates;
		Added extra verification during password-resets;
		Fixed a password-reset bug;
		Added some eyecandy during image-loads;
		Added a progressindicator to commandline imports;
		Added a few LightBoxes where appropriate;
		Database renamed, setupscript rewritten;
		Updated list of release-dates (as of 2011-07-15);
		Lots of code clean-up and minor cosmetic surgery; 

2011-07-04 1.0
		First public release.



CONTACT:
--------
Name:	FWieP (main developer)
Email:	fwiep@fwiep.nl
WWW:	https://code.google.com/p/candydolldb/
SVN:	http://candydolldb.googlecode.com/svn/trunk/


History.md
# History #

Once upon a time, in a land perhaps closer than you think, there was a programmer. He regularly thought up projects in PHP and MySQL which would force him to familiarize himself with all the ins and outs he thought existed. Boy, was he in for a surprise... At work, he programmed in C# (.NET framework), strictly [Object Oriented](http://en.wikipedia.org/wiki/Object-oriented_programming) - could this be done in PHP, too?

After about two weeks of thinking it over, he needed a subject around which to build his new project. CandyDoll sprang to mind. After searching the internet for an existing application, and finding none, he decided [CandyDollDB](http://code.google.com/p/candydolldb/) had to be built. The preset goal would be to comfortably manage that vast collection of CandyDoll images and videos through a web-interface.

Then finally, one sunny sunday morning, the programmer thought of Google Code. He decided he would start a project for his application. This was (and is) a completely new learning-curve. How does one write comprehensible documentation? How does one anticipate the users' needs and wants? Are there actually any people out there who would benefit from this application?

I hope, the answer to that last question is YES. The fact that you are reading this page right now proves just that. :-)

Of course, collecting the actual content is as much fun as programming the interface: Browsing usenet, forums and the official [CandyDoll](http://www.candydoll.tv/)-site for that latest (or lost) bit of info about models, sets or collections.

Please enjoy CandyDollDB and your CandyDoll-collection, see you soon.
<br>
<br>
<br>
<i><b>FWieP</b></i>
<br><i>main developer CandyDollDB</i>

cd.php
<?php

error_reporting(E_ALL);
ini_set('display_errors', '1');
date_default_timezone_set(@date_default_timezone_get());

include('class/class.i18n.php');
include('class/class.i18n.en.php');
include('class/class.i18n.nl.php');
include('class/class.i18n.de.php');

$lang = new i18n();

if(isset($argv) && $argc > 0)
{
	$configPath = sprintf('%1$s/config.php', dirname($_SERVER['PHP_SELF']));
	if(file_exists($configPath))
	{ require_once($configPath); }
}
else
{
	if(file_exists('config.php'))
	{ require_once('config.php'); }
}

if(!defined('DBHOSTNAME') || strlen(DBHOSTNAME) == 0 ||
   !defined('DBUSERNAME') || strlen(DBUSERNAME) == 0 ||
   !defined('DBPASSWORD'))
{
	if(!array_key_exists('REQUEST_URI', $_SERVER))
	{
		echo $lang->g('ErrorPleaseUseWebInterfaceForSetup')."\n";
		exit(128);
	}
	else if(stristr(basename($_SERVER['REQUEST_URI']), 'setup') === FALSE)
	{
		header('location:setup.php');
		exit(128);
	}
}

if(!defined('DBNAME'))
{ define('DBNAME', 'candydolldb'); }

define('CANDYDOLLDB_VERSION', '1.9');

/* Rights */
	define('RIGHT_ACCOUNT_LOGIN',	1);
	define('RIGHT_ACCOUNT_EDIT',	2);
	define('RIGHT_ACCOUNT_PASSWORD',3);
	
	define('RIGHT_MODEL_ADD',		4);
	define('RIGHT_MODEL_EDIT',		5);
	define('RIGHT_MODEL_DELETE',	6);
	
	define('RIGHT_SET_ADD',			7);
	define('RIGHT_SET_EDIT',		8);
	define('RIGHT_SET_DELETE',		9);
	
	define('RIGHT_IMAGE_ADD',		10);
	define('RIGHT_IMAGE_EDIT',		11);
	define('RIGHT_IMAGE_DELETE',	12);
	
	define('RIGHT_VIDEO_ADD',		13);
	define('RIGHT_VIDEO_EDIT',		14);
	define('RIGHT_VIDEO_DELETE',	15);
	
	define('RIGHT_TAG_ADD',			16);
	define('RIGHT_TAG_EDIT',		17);
	define('RIGHT_TAG_DELETE',		18);
	
	define('RIGHT_USER_ADD',		19);
	define('RIGHT_USER_EDIT',		20);
	define('RIGHT_USER_DELETE',		21);
	define('RIGHT_USER_RIGHTS',		22);
	
	define('RIGHT_IMPORT_XML',		23);
	define('RIGHT_EXPORT_XML',		24);
	
	define('RIGHT_EXPORT_ZIP',		25);
	define('RIGHT_EXPORT_ZIP_MULTI',26);
	
	define('RIGHT_EXPORT_SFV',		27);
	define('RIGHT_EXPORT_CSV',		28);
	define('RIGHT_EXPORT_INDEX',	29);
	define('RIGHT_EXPORT_VIDEO',	30);
	
	define('RIGHT_SEARCH_DIRTY',	31);
	define('RIGHT_SEARCH_TAGS',		32);
	
	define('RIGHT_CACHE_CLEANUP',	33);
	define('RIGHT_TAG_CLEANUP',		34);
	define('RIGHT_CACHE_DELETE',	35);
	define('RIGHT_CONFIG_REWRITE',	36);
/* /Rights */
	
define('GENDER_UNKNOWN', 0);
define('GENDER_FEMALE', 1);
define('GENDER_MALE', 2);

define('SET_CONTENT_NONE',  0);
define('SET_CONTENT_IMAGE', 1);
define('SET_CONTENT_VIDEO', 2);

define('DATE_KIND_UNKNOWN',  0);
define('DATE_KIND_IMAGE', 1);
define('DATE_KIND_VIDEO', 2);

define('EXPORT_PATH_OPTION_NONE', 0);
define('EXPORT_PATH_OPTION_RELATIVE', 1);
define('EXPORT_PATH_OPTION_FULL',  2);

define('EXPORT_OPTION_SERIALIZE', 0);
define('EXPORT_OPTION_IMAGE', 1);
define('EXPORT_OPTION_ZIP', 2);

define('CACHEIMAGE_KIND_UNKNOWN', 0);
define('CACHEIMAGE_KIND_IMAGE', 1);
define('CACHEIMAGE_KIND_VIDEO', 2);
define('CACHEIMAGE_KIND_SET',	4);
define('CACHEIMAGE_KIND_INDEX', 8);
define('CACHEIMAGE_KIND_MODEL', 16);

define('COMMAND_DELETE', 'del');

define('LOGIN_ERR_PASSWORDSNOTIDENTICAL', 33362);
define('LOGIN_ERR_USERNAMEANDMAILADDRESNOTFOUND', 33369);
define('LOGIN_ERR_RESETCODENOTFOUND', 33363);
define('LOGIN_ERR_USERNAMENOTFOUND', 33364);
define('LOGIN_ERR_PASSWORDINCORRECT', 33365);
define('SQL_ERR_NOSUCHTABLE', 33366);
define('SYNTAX_ERR_EMAILADDRESS', 33367);
define('REQUIRED_FIELD_MISSING', 33368);
define('XML_ERR_XML_VALID', 33369);
define('XML_ERR_SCHEMA_VALID', 33370);
define('RIGHTS_ERR_USERNOTALLOWED', 33371);

$CSVRegex = '/\s*(?<!\\\),\s*/';
$DateStyleArray = array(
	"d-m-Y",	// 14-04-2012
	"j-n-Y",	// 14-4-2012
	"j F Y",	// 14 April 2012 
	"m-d-Y",	// 04-14-2012 
	"n-j-Y",	// 4-14-2012 
	"F j, Y"	// April 14, 2012 
);

include('class/class.global.php');
include('class/class.error.php');
include('class/class.info.php');
include('class/class.html.php');
include('class/class.dbi.php');

@session_start();

if(!array_key_exists('Errors', $_SESSION))
{ $_SESSION['Errors'] = serialize(array()); }

if(!array_key_exists('Infos', $_SESSION))
{ $_SESSION['Infos'] = serialize(array()); }

if(defined('DBHOSTNAME') &&
   defined('DBUSERNAME') &&
   defined('DBPASSWORD'))
{
	$dbi = @new DBi(DBHOSTNAME, DBUSERNAME, DBPASSWORD, DBNAME);
	if(mysqli_connect_errno() == 0)
	{
		$dbi->query("SET GLOBAL sql_mode = 'STRICT_ALL_TABLES';");
	}
	else
	{
		echo $lang->g('ErrorSetupConnectDatabase');
		exit;
	}
}

include('class/class.user.php');
include('class/class.date.php');
include('class/class.image.php');
include('class/class.cacheimage.php');
include('class/class.model.php');
include('class/class.set.php');
include('class/class.video.php');
include('class/class.tag.php');
include('class/class.tag2all.php');
include('class/class.indexgenerator.php');

$EmailPages = array('password.php');
if(in_array(basename($_SERVER['PHP_SELF']), $EmailPages))
{
	require 'class/class.phpmailer.php';
	$ml = new PHPMailer();
	$ml->IsSMTP();
	$ml->From = SMTP_FROM_ADDRESS;
	$ml->FromName = SMTP_FROM_NAME;
	$ml->Host = SMTP_HOST;
	$ml->Username = SMTP_USERNAME;
	$ml->Password = SMTP_PASSWORD;
	$ml->Port = SMTP_PORT;
	$ml->SMTPAuth = SMTP_AUTH;
}

?>
